### PR TITLE
Incremental builder

### LINKED ISSUE #653: ixbrl-viewer multiple POST thinks its a document set
### What happened?

We're running into issues using the ixbrl-viewer for ESMA filings. We're using web server. When we run a package for company #1 using post we get correct results with a single correct output file that works correctly. However, when we run a second, different package, for company #2 we get incorrect output files. Two different html files are output. When we open the first file we get the following message:

"Displaying iXBRL document sets from local files is not supported. Please view the viewer files using a web server."

It appears to think it's a document set when the second post is sent. Is there a way to let the plug-in know that this is a different filing with a single instance and not a document set?

Below is what is being sent:
image

Austin Matherne emailed me about this and said, "I suspect this was caused by the changes in the ixbrl-viewer to add multi ixds support a little while back. Would you mind opening an issue in the for @paulwarren-wk to take a look at?"

Thanks as always,
Chad

### Version

Current master

### With which browsers are you experiencing the bug?

Chrome

### Documents

_No response_

### Screenshots

While doing the investigation of the iXBRL viewer integration possibilities, an obstacle has been encountered. The current Arelle implementation leverages the usage of a web server to handle client requests. Upon exploring the potential ways of adding the new plugin to the existing infrastructure, it was noticed after making a second request using a different ESEF package obtained here https://filings.xbrl.org/ to the /validation endpoint, two HTML files are added to the output zip. One is the instance file of the for the company provided in the payload which is without an iXBRL viewer augmentation and the other one is with an iXBRL viewer augmentation, but an alert is shown when opening it and it does not seem to load. The first request provides the desired result - an HTML file enhanced with the iXBRL viewer capabilities.

Command and arguments used to launch the web server with the iXBRL viewer plugin in Arelle, the current latest viewer version is specified:

image

Results after pinging the /validation?media=zip endpoint with an ESEF package payload. The results are as expected:

One HTML file:

image

iXBRL viewer can be seen upon opening the HTML file:
image

Results after firing a second request to the same endpoint with a different ESEF package payload. In the output zip there are two HTML files, neither of which successfully present a fully loaded iXBRL viewer upon opening. This behavior is undesired:

Two HTML files in the output package:
image

HTML based on the package of the payload in the first request, viewer is added but does not load:
image

HTML based on the payload package in the second request, no enhancement with the iXBRL viewer is to be seen:
Image
image

Taking such results into account, we are faced with an integration difficulty, since the current infrastructure of handling Arelle related requests is based on launching a configured web server.

### PR DESCRIPTION
**Pull request by paulwarren-wk**
#### Reason for change

Fixes #653.  Avoids retaining loaded XBRL models indefinitely.

#### Description of change

Previously the viewer used the `keepOpen` feature in Arelle to retain XBRL models after the are loaded, so that multiple files can be assembled into a single viewer.  This change uses the new `pluginData` feature to preserve viewer information across models when building multi-document viewers, and then creating the viewer in a separate call after all models are processed.

A slightly awkward feature of this is that we need to retain the XML models until we serialize at the end, but the XML models created by Arelle use custom element classes that are no longer safe to use once the Arelle models have been disposed.  I have fixed this by unsetting the custom element class lookup after taking a copy of the model. 

#### Steps to Test

See steps on #653.  Can also be confirmed by submitting the same file to the web service multiple times, and confirming that you get exactly the same file back each time.

**review**:
@Arelle/arelle
@paulwarren-wk

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: iXBRLViewerPlugin/iXBRLViewer.py
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py:516]: > Out of curiosity, why is this needed for serialization? I would have guessed that parser.set_element_class_lookup is only used when parsing.

That would have been my expectation too.  It's a call to `getroot()` that triggers it:

```
Exception 'ModelXbrl' object has no attribute 'qnameConcepts'
...
    for child in viewerFile.xmlDocument.getroot():
   File "src/lxml/etree.pyx", line 1219, in lxml.etree._Element.__iter__
   File "src/lxml/etree.pyx", line 2877, in lxml.etree.ElementChildIterator.__cinit__
   File "src/lxml/etree.pyx", line 1636, in lxml.etree._elementFactory
   File "src/lxml/classlookup.pxi", line 403, in lxml.etree._parser_class_lookup
   File "src/lxml/classlookup.pxi", line 456, in lxml.etree._custom_class_lookup
   File "/home/pdw/Arelle/arelle/ModelObjectFactory.py", line 141, in lookup
    return self.modelXbrl.matchSubstitutionGroup(qnameNsLocalName(ns, ln), elementSubstitutionModelClass)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   File "/home/pdw/Arelle/arelle/ModelXbrl.py", line 471, in matchSubstitutionGroup
    elementMdlObj = self.qnameConcepts.get(elementQname)
                    ^^^^^^^^^^^^^^^^^^
```

Looking at the [etree source](https://github.com/lxml/lxml/blob/lxml-5.1/src/lxml/etree.pyx#L2900) I think it's creating Python proxy nodes on-the-fly when needed based on the C nodes (see also [proxy.pxi](https://github.com/lxml/lxml/blob/lxml-5.1/src/lxml/proxy.pxi)), so depending on what's cached, an element class lookup can happen when reading nodes.

> what code is doing the clean-up, and what is it doing? 

I think it's [ModelXbrl.close()](https://github.com/Arelle/Arelle/blob/master/arelle/ModelXbrl.py#L369) and I think it's `self.__dict__.clear()` that kills the `qnameConcepts` dict, which is used when trying to lookup substitution groups when instantiating Python proxies.

> Does "not safe" just mean that the code fails with a NoneError (or whatever)

Yes - the above Python exception.

Having now looked more carefully at the etree/libxml2 source code, I think that `deepcopy` only copies the C nodes and this doesn't include the `_private` value where etree stores its proxy nodes, so doing the deepcopy clears the cache of Python proxy nodes, and once the custom class lookup is removed we're left with a clean lxml.etree tree.

Comment (brettkail-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py:524]: Thanks, that helps.  Maybe this comment is clearer?
> deepcopy does not retain the Python proxies, so iterating the node tree must create new ones.  However, the original ModelObjectFactory is still referenced, and that references a potentially closed ModelXbrl, so we must use a different lookup object.  We don't rely on any specific proxy, so just use the default.

That said, if I'm reading the code correctly, it seems like even with this change, we're still using the [stale element factory](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L818-L839) for the root element and just getting lucky that the Arelle logic doesn't fail for it?  I guess we could extend the comment to say that.  Alternatively, I guess we could serialize/reparse using etree (unfortunate), or we could avoid deepcopy and manually copy into a custom mini-etree that only supports the features needed by XHTMLSerializer...

Comment (paulwarren-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py:524]: I don't understand what circumstances the [`if root is not None`](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L831) line applies to, but yes, it does seem possible that we're using the Arelle class lookup.

Other options:

1. Have Arelle unset the class lookup on clean up.
2. Detect if the models have been cleaned up, and if so, unset the class lookup before the `deepcopy`
3. Stash and restore the class lookup before/after the `deepcopy`, but this relies on reading the private `_class_lookup` member.

Comment (paulwarren-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py:528]: > That said, if I'm reading the code correctly, it seems like even with this change, we're still using the [stale element factory](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L818-L839) for the root element and just getting lucky that the Arelle logic doesn't fail for it?

Having looked at this again, I think this is actually safe. The element factory isn't stale at the point that we do the copy, because we instantiate the `iXBRLViewerFile` object as part of processing the model.  

The serialization happens later, but by that time we've replaced the class lookup.  It does mean that we potentially serialize an Arelle proxy object for the root, but I think that's safe as the Arelle proxies don't interfere with standard lxml operations.

I've revised the comment in cdd035e

Comment (brettkail-wk) [reply iXBRLViewerPlugin/iXBRLViewer.py:528]: That makes sense, thanks.

**Comment by paulwarren-wk** [comment]
I'm afraid I'm at a loss with the remaining [type linting error](https://github.com/Arelle/ixbrl-viewer/actions/runs/9603850665/job/26488006286?pr=678#step:5:61)

It seems like `mypy` is failing to find the definition of `PluginData`.

**Comment by paulwarren-wk** [comment]
> Arelle is not considered a typed library

Ah, thank you. I'd assumed that as we reference Arelle types  elsewhere that we were making use of Arelle type information, e.g.:

```python
    def addFact(self, report: ModelXbrl, f):
```

But I assume that sub-classing is a special case where the type information is actually needed?

OOI, what is needed for Arelle to be considered a typed library? Is it just missing type annotations, or is there some package configuration needed?

**Comment by austinmatherne-wk** [comment]
@paulwarren-wk The `# type: ignore[misc]` option Steven suggested is the best path forward at the moment. We plan to publish the Arelle types, but need to sort out a few internal details first.

**Comment by paulwarren-wk** [comment]
> We'd need to choose one of [the options](https://typing.readthedocs.io/en/latest/spec/distributing.html#packaging-typed-libraries) to support typing. Alternatively, mypy might implement a [feature](https://www.github.com/python/mypy/issues/8545) to consider a module as providing types, even if it's not officially supported.

Thank you - that's much clearer now.  

I've now added the comment to ignore type checks on this line (d5cccaa)

**Review by austinmatherne-wk — DISMISSED** [review]

File: iXBRLViewerPlugin/__init__.py
Comment (austinmatherne-wk) [inline iXBRLViewerPlugin/__init__.py:323]: This is missing the cntlr parameter.

When run from the GUI:
```shell
[viewer:exception] Exception processModel() missing 1 required positional argument: 'modelXbrl' 
Traceback ['  File "/Users/austinmatherne/git/Arelle/arelle/plugin/iXBRLViewerPlugin/__init__.py", line 323, in guiRun\n    processModel(modelXbrl)\n'] - wk-20230930.htm 
```

**Comment by michaelchadfleming** [comment]
@paulwarren-wk @austinmatherne-wk Hi Guys, I hope you had an amazing weekend! :) This is in regards to the comment on https://github.com/Arelle/ixbrl-viewer/issues/653#ref-pullrequest-2305424002. Does this pull request resolve the bug? It appears that there were conflicts with the pull? Thank you for your valuable time, Chad

File: iXBRLViewerPlugin/__init__.py
Comment (paulwarren-wk) [reply iXBRLViewerPlugin/__init__.py]: Fixed in [36f1c09](https://github.com/Arelle/ixbrl-viewer/commit/36f1c09ac65b05db24ff89e8bec1062060d1882d)

**Comment by paulwarren-wk** [comment]
> @paulwarren-wk @austinmatherne-wk Hi Guys, I hope you had an amazing weekend! :) This is in regards to the comment on [#653 (reference)](https://github.com/Arelle/ixbrl-viewer/issues/653#ref-pullrequest-2305424002). Does this pull request resolve the bug? It appears that there were conflicts with the pull? Thank you for your valuable time, Chad

Sorry @michaelchadfleming, I dropped the ball on this and didn't realise it was blocked on me.  I've just pushed a fix for what I think is the final issue.

**Review by austinmatherne-wk — APPROVED** [review]

**Review by derekgengenbacher-wf — APPROVED** [review]

**Pull request merged**
